src/backend/input.rs

```diff
@@ -1,31 +1,59 @@
 use std::hash::Hash;
 
-#[derive(Hash, PartialEq, Eq, Clone, Copy)]
-pub enum InputDevice<T: PartialEq + Eq + Hash + Clone + Copy>
+pub type Seat = u32;
+
+#[derive(Debug, PartialEq, Eq, Clone, Copy)]
+pub enum KeyState {
+    Released,
+    Pressed,
+}
+
+#[derive(Debug, PartialEq, Eq, Clone, Copy)]
+pub enum MouseButton {
+    Left,
+    Middle,
+    Right,
+    Other(u8),
+}
+
+#[derive(Debug, PartialEq, Eq, Clone, Copy)]
+pub enum MouseButtonState {
+    Released,
+    Pressed,
+}
+
+#[derive(Debug, PartialEq, Eq, Clone, Copy)]
+pub enum Axis
+{
+    Vertical,
+    Horizontal,
+}
+
+#[derive(Debug, PartialEq, Eq, Clone, Copy)]
+pub enum AxisSource
+{
+    Finger,
+    Continous,
+    Wheel,
+    WheelTilt,
+}
+
+#[derive(Debug, PartialEq, Clone, Copy)]
+pub enum TouchEvent
 {
-    Keyboard(Keyboard<T>),
-    Mouse(Mouse<T>),
-    Touchscreen(Touchscreen<T>),
-    // What about Gamepads in Wayland?
-    // Anything else?
+    Down { x: f64, y: f64 },
+    Motion { x: f64, y: f64 },
+    Up,
+    Cancel,
+    Frame
 }
 
-//TODO
-pub type ModifierMask = ();
-
-#[derive(Hash, PartialEq, Eq, Clone, Copy)]
-pub struct Keyboard<T: PartialEq + Eq + Hash + Clone + Copy> { id: T }
-#[derive(Hash, PartialEq, Eq, Clone, Copy)]
-pub struct Mouse<T: PartialEq + Eq + Hash + Clone + Copy> { id: T }
-#[derive(Hash, PartialEq, Eq, Clone, Copy)]
-pub struct Touchscreen<T: PartialEq + Eq + Hash + Clone + Copy> { id: T }
-
-trait InputBackendHandler<T: PartialEq + Eq + Hash + Clone + Copy> {
-    fn on_input_attached(&mut self, device: InputDevice<T>);
-    fn on_input_detached(&mut self, device: InputDevice<T>);
-    fn on_kbd_event(&mut self, device: Keyboard<T>, modifiers: ModifierMask, event: ()); //TODO discussion required
-    fn on_mouse_move(&mut self, device: Mouse<T>, to: (u32, u32));
-    fn on_mouse_button(&mut self, device: Mouse<T>, modifiers: ModifierMask, button: ()/*TODO*/, state: ()/*TODO: Pressed or Not*/);
-    fn on_mouse_scroll(&mut self, device: Mouse<T>, modifiers: ModifierMask, by: i32, axis: ()/*TODO*/);
-    fn on_touch(&mut self, device: Touchscreen<T>, modifiers: ModifierMask, at: (u32, u32)); //TODO this is most likely to raw. We should aggregate touch events
+pub trait InputBackendHandler {
+    fn on_seat_created(&mut self, seat: &Seat);
+    fn on_seat_destroyed(&mut self, seat: &Seat);
+    fn on_keyboard_key(&mut self, seat: &Seat, time: u32, key_code: u32, state: KeyState, count: u32);
+    fn on_pointer_move(&mut self, seat: &Seat, time: u32, to: (u32, u32));
+    fn on_pointer_button(&mut self, seat: &Seat, time: u32, button: MouseButton, state: MouseButtonState);
+    fn on_pointer_scroll(&mut self, seat: &Seat, time: u32, axis: Axis, source: AxisSource, amount: f64);
+    fn on_touch(&mut self, seat: &Seat, time: u32, event: TouchEvent);
 }
```

Comment:
We need a callback to advertise the capabilities of a seat (keyboard/mouse/touch), the wayland protocol requires this.